### PR TITLE
parser: only allow int column in range partition

### PR DESCRIPTION
**Pull request by zhexuany**
For the following query:

```
CREATE TABLE members ( firstname VARCHAR(25) NOT NULL, lastname VARCHAR(25) NOT NULL, username VARCHAR(16) NOT NULL, email VARCHAR(35), joined DATE NOT NULL)PARTITION BY RANGE( joined ) ( PARTITION p0 VALUES LESS THAN (1960), PARTITION p1 VALUES LESS THAN (1970), PARTITION p2 VALUES LESS THAN (1980), PARTITION p3 VALUES LESS THAN (1990), PARTITION p4 VALUES LESS THAN MAXVALUE);
```

mysql return:

```
ERROR 1105 (HY000): Field '`joined`' is of a not allowed type for this type of partitioning
```

This PR aims to fix this inconsistent behavior.

### CODE REVIEW

**Comment by zhexuany** [comment]
@coocood @jackysp PTAL

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -806,6 +806,14 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 			buf := new(bytes.Buffer)
 			s.Partition.Expr.Format(buf)
 			pi.Expr = buf.String()
+			for _, col := range colDefs {
+				if col.Name.Name.L == pi.Expr && col.Tp.Tp != types.KindInt64 {
+					return errors.Trace(fmt.Errorf("Field '%s' is of a not allowed type for this type of partitioning", pi.Expr))
+				}
+			}
+			// fmt.Printf("type is %d", s.Partition.Expr.GetType().Tp)
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Remove debug log

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -806,6 +806,14 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 			buf := new(bytes.Buffer)
 			s.Partition.Expr.Format(buf)
 			pi.Expr = buf.String()
+			for _, col := range colDefs {
+				if col.Name.Name.L == pi.Expr && col.Tp.Tp != types.KindInt64 {
+					return errors.Trace(fmt.Errorf("Field '%s' is of a not allowed type for this type of partitioning", pi.Expr))
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: And define mysql error code for the error

**Comment by zhexuany** [comment]
@coocood @jackysp PTAL. This PR fixes `ColumnNameExpr` case. For `FuncCallExpr`, I will create another PR to fix it.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -806,6 +806,16 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 			buf := new(bytes.Buffer)
 			s.Partition.Expr.Format(buf)
 			pi.Expr = buf.String()
+			if s.Partition.Tp == model.PartitionTypeRange {
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					for _, col := range cols {
+						name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+						if fmt.Sprintf("`%s`", name) == pi.Expr && col.Tp != mysql.TypeLong {
+							return errors.Trace(fmt.Errorf("Field '%s' is of a not allowed type for this type of partitioning", pi.Expr))
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: There is a example on how to do it:
https://github.com/pingcap/tidb/blob/4b2b4bed2cafc983cee2fd754b607788e15306df/ddl/ddl.go#L574

Error code is important information, that's why we need to define them for MySQL errors.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -806,6 +806,16 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 			buf := new(bytes.Buffer)
 			s.Partition.Expr.Format(buf)
 			pi.Expr = buf.String()
+			if s.Partition.Tp == model.PartitionTypeRange {
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					for _, col := range cols {
+						name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+						if fmt.Sprintf("`%s`", name) == pi.Expr && col.Tp != mysql.TypeLong {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Although it doesn't matter here, I'd like to point out that this is better for performance:
```if col.Tp != mysql.TypeLong && fmt.Sprintf("`%s`", name) == pi.Expr {```

File: executor/show_test.go [hunk 5]
```diff
@@ -324,6 +324,13 @@ func (s *testSuite) TestShow(c *C) {
 			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"+"\nPARTITION BY RANGE ( `a` ) (\n  PARTITION p0 VALUES LESS THAN 10,\n  PARTITION p1 VALUES LESS THAN 20,\n  PARTITION p2 VALUES LESS THAN MAXVALUE\n)",
 	))
 
+	tk.MustExec(`drop table if exists t`)
+	_, err := tk.Exec(`CREATE TABLE t (x int, y char) PARTITION BY RANGE(y) (
```
Comment (tiancaiamao) [inline executor/show_test.go]: Good job!

File: ddl/ddl_db_test.go [hunk 6]
```diff
@@ -162,8 +162,6 @@ func (s *testDBSuite) TestMySQLErrorCode(c *C) {
 	s.testErrorCode(c, sql, tmysql.ErrUnknownCharacterSet)
 	sql = "create table t1(a int) character set laitn1;"
 	s.testErrorCode(c, sql, tmysql.ErrUnknownCharacterSet)
-	sql = "create table t2(c1.c2 blob default null);"
-	s.testErrorCode(c, sql, tmysql.ErrWrongTableName)
```
Comment (jackysp) [inline ddl/ddl_db_test.go]: Why remove this case?

Comment (zhexuany) [reply ddl/ddl_db_test.go]: Added them back. They were accidentally deleted from codebase and nothing to do with this PR.

File: ddl/ddl_api.go (hunk 4 above)
Comment (zhexuany) [reply ddl/ddl_api.go]: DONE

File: ddl/ddl_api.go (hunk 3 above)
Comment (zhexuany) [reply ddl/ddl_api.go]: DONE

**Comment by zhexuany** [comment]
@jackysp @tiancaiamao PTAL again.

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -806,6 +806,16 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 			buf := new(bytes.Buffer)
 			s.Partition.Expr.Format(buf)
 			pi.Expr = buf.String()
+			if s.Partition.Tp == model.PartitionTypeRange {
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					for _, col := range cols {
+						name := strings.Replace(col.Name.String(), ".", "`.`", -1)
```
Comment (lamxTyler) [inline ddl/ddl_api.go]: What about `TypeLonglong`?

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -806,6 +806,16 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
+						if col.Tp != mysql.TypeLong && fmt.Sprintf("`%s`", name) == pi.Expr {
```
Comment (lamxTyler) [inline ddl/ddl_api.go]: What about TypeLonglong?

**Comment by jackysp** [comment]
Please fix CI first.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**